### ISSUE TITLE
inconsistent type error reporting - list invariance, literal inference

### ISSUE THREAD

**Issue by beauxq** [microsoft/pyright#5807]
**Describe the bug**

```python3
def make_something_a() -> list[tuple[str]]:
    x = [("a",)]
    return x


def make_something_b() -> list[tuple[str]]:
    return [("a",)]
```

In `make_something_a`, it's reported that I'm not returning the correct type:
> Expression of type "list[tuple[Literal['a']]]" cannot be assigned to return type "list[tuple[str]]"
>   "list[tuple[Literal['a']]]" is incompatible with "list[tuple[str]]"
>     Type parameter "_T@list" is invariant, but "tuple[Literal['a']]" is not the same as "tuple[str]"

`make_something_b` doesn't report any problem.

They're both returning the same thing.

**VS Code extension or command-line**
Pylance v2023.8.40

**Comment by erictraut** [comment]
This behavior is correct. The difference is that the first example is using default type inference rules for lists and tuples. The second example is using bidirectional type inference to override the normal default rules. If you want to get the same results, you need to provide a type annotation for variable `x` in the first example.

```python
def make_something_a() -> list[tuple[str]]:
    x: list[tuple[str]] = [("a",)]
    return x
```

For more details about the default type inference rules for [lists](https://microsoft.github.io/pyright/#/type-inference?id=list-expressions) and [https://microsoft.github.io/pyright/#/type-inference?id=tuple-expressions](tuples and information about [bidirectional type inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional)-type-inference-expected-types), please refer to the pyright documentation.
